File: Python/LeetCodeTest/lc056_MergeIntervals.py
from typing import List

class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]: # 先依照每个段落的开始值排序，然后只需要比较结束值
        intervals.sort(key=lambda x: x[0])
        output = []
        for interval in intervals:
            if len(output) == 0 or output[-1][1] < interval[0]:
                output.append(interval)
            else:
                output[-1][1] = max(output[-1][1], interval[1])
        return output

    # 把每个数字转换为线段上的合并，那么就包含数字点和数字点接下来的线段，所以存储空间使用2倍，
    # 这种方式是可行的，但会占用很多空间，而且并不高效
    def merge_inefficient(self, intervals: List[List[int]]) -> List[List[int]]:
        max_cache = 0
        for interval in intervals:
            for data in interval:
                max_cache = max(data, max_cache)
        cache = [False] * (max_cache + 1) * 2 # 分别代表等于i和i到i+1是否包含，用于解决半开半闭
        for interval in intervals:
            for data in range(interval[0] * 2, interval[1] * 2):
                cache[data] = True
            if interval[0] == interval[1]:
                cache[interval[0] * 2] = True
        output = []
        start = 0
        found = False
        for i in range(len(cache)):
            data = cache[i]
            if data and not found:
                start = i
                found = True
            if not data and found:
                output.append([start // 2, i // 2])
                found = False
        return output

    def merge_error(self, intervals: List[List[int]]) -> List[List[int]]: # 无法解决形如[[1,4],[0,0]]的输入
        max_cache = 0
        for interval in intervals:
            for data in interval:
                max_cache = max(data, max_cache)
        cache = [False] * (max_cache + 1)
        for interval in intervals:
            for data in range(interval[0], interval[1]):
                cache[data] = True
        output = []
        start = 0
        found = False
        for i in range(len(cache)):
            data = cache[i]
            if data and not found:
                start = i
                found = True
            if not data and found:
                output.append([start, i])
                found = False
        return output

if __name__ == "__main__":
    solution = Solution()
    result = solution.merge([[1,4],[0,0]])
    print(result)
    result = solution.merge([[2,4],[1,1]])
    print(result)
    result = solution.merge([[1,3],[2,6],[8,10],[15,18]])
    print(result)

File: Java/src/com/liu/LeetCode/FifteenPatterns/OverlappingIntervals/LC056_MergeIntervals.java
package com.liu.LeetCode.FifteenPatterns.OverlappingIntervals;

import java.util.*;

/**
 * The Overlapping Intervals pattern is used to merge or handle overlapping intervals in an array.
 * In an interval array sorted by start time, two intervals [a, b] and [c, d] overlap if b >= c (i.e., the end time of the first interval is greater than or equal to the start time of the second interval).
 * Sample Problem:
 * Problem Statement: Merge all overlapping intervals.
 * Example:
 * Input: intervals = [[1, 3], [2, 6], [8, 10], [15, 18]]
 * Output: [[1, 6], [8, 10], [15, 18]]
 * Explanation:
 * Sort the intervals by their start time.
 * Create an empty list called merged to store the merged intervals.
 * Iterate through the intervals and check if it overlaps with the last interval in the merged list.
 * If it overlaps, merge the intervals by updating the end time of the last interval in merged.
 * If it does not overlap, simply add the current interval to the merged list.
 */

public class LC056_MergeIntervals {

    private static int[][] mergeMy1(int[][] intervals) { // accepted 但性能相当差
        int length = intervals.length;
        if (length < 2) {
            return intervals;
        }
        List<int[]> cacheList = new ArrayList<>();
        Collections.addAll(cacheList, intervals);
        cacheList.sort(Comparator.comparingInt(o -> o[0]));
        int index = 1;
        while (index < cacheList.size()) {
            int[] lastInterval = cacheList.get(index - 1);
            int[] interval = cacheList.get(index);
            if (interval[0] <= lastInterval[1]) {
                lastInterval[1] = Math.max(lastInterval[1], interval[1]);
                cacheList.remove(interval);
            } else {
                index++;
            }
        }
        return cacheList.toArray(new int[cacheList.size()][2]);
    }

    private static int[][] mergeAnswer1(int[][] intervals) {
        Arrays.sort(intervals, Comparator.comparingInt(a -> a[0]));

        List<int[]> merged = new ArrayList<>();
        int[] previous = intervals[0];

        for (int i = 1; i < intervals.length; i++) {
            int[] interval = intervals[i];
            if (interval[0] <= previous[1]) {
                previous[1] = Math.max(previous[1], interval[1]);
            } else {
                merged.add(previous);
                previous = interval;
            }
        }

        merged.add(previous);

        return merged.toArray(new int[merged.size()][]);
    }

    public static void main(String[] args) {
        int[][][] intervalsSet = new int[][][]{
                {{1, 3}, {2, 6}, {5, 10}, {8, 18}},
                {{1, 3}},
                {{1, 3}, {2, 6}, {8, 10}, {15, 18}},
                {{1, 4}, {4, 5}},
        };
        int[][] result;
        long startTime;
        long endTime;

        for (int i = 0; i < intervalsSet.length; i++) {
            startTime = System.currentTimeMillis();
            result = mergeMy1(intervalsSet[i]);
            endTime = System.currentTimeMillis();
            System.out.print("merge My1 result" + i + " ");
            for (int[] list : result) {
                System.out.print("[");
                for (int num : list) {
                    System.out.print(num + ",");
                }
                System.out.print("]; ");
            }
            System.out.println(" during time " + (endTime - startTime));
        }

        for (int i = 0; i < intervalsSet.length; i++) {
            startTime = System.currentTimeMillis();
            result = mergeAnswer1(intervalsSet[i]);
            endTime = System.currentTimeMillis();
            System.out.print("merge Answer1 result" + i + " ");
            for (int[] list : result) {
                System.out.print("[");
                for (int num : list) {
                    System.out.print(num + ",");
                }
                System.out.print("]; ");
            }
            System.out.println(" during time " + (endTime - startTime));
        }
    }
}
